File: tests/cpu/seats/seat-rows.test.ts
import { describe, it, expect } from 'vitest';
import { makeMap, makeState, tileAtCoords, settleAt, grantCivics } from '../helpers';
import { emptySeat, setTileOwner } from '../../../cpu/core/seats';
import { computeCityStats } from '../../../cpu/core/city';
import { getModifiers, wonderExtraSlots } from '../../../cpu/core/effects';
import { greatPersonPointsPerTurn } from '../../../cpu/core/greatPeople';
import { unitKillEvent } from '../../../cpu/core/eras';
import { rosterCS } from '../../../cpu/core/combat';
import { spawnUnit } from '../../../cpu/core/units';
import { districtAdjacency } from '../../../cpu/core/yields';
import { neighbors } from '../../../world/hex';
import { CIV_LEADERS } from '../../../cpu/data/seats';
import { HAPPY_YIELD_ROWS, HAPPY_GPP_ROWS, POLICY_SLOT_ROWS, POST_COMBAT_YIELD_ROWS, DISTRICT_ADJ_ROWS } from '../../../cpu/data/civilizations';
import { UNITS } from '../../../cpu/data/units';
import { RESOURCES } from '../../../world/resources';
import type { City, GameState, Tile } from '../../../cpu/core/types';

/**
 * THE SEAT'S ROSTER ROWS (CIV6, the install's TraitModifiers): the happiness
 * percentages and Great Person points of the Scottish Enlightenment, the
 * government slot of Plato's Republic and the Holy Roman Emperor, the Culture
 * and Faith a kill pays Gorgo and Tamar, Thermopylae's per-policy strength,
 * and the Amazon's rainforest adjacency.
 */
const seatRow = (civ: string) => CIV_LEADERS.findIndex((l) => l.civ === civ);
const leaderRow = (leader: string) => CIV_LEADERS.findIndex((l) => l.leader === leader);

function sceneAs(row: number): GameState {
  const state = makeState(makeMap(14, 14, 'GRASSLAND'));
  state.seats.push(emptySeat(1));
  state.seats[0].civ = row;
  state.seats[1].civ = seatRow('AMERICA');
  return state;
}

/** Improved luxuries this seat owns — each is one Amenity in its four
 *  nearest cities, which is how a scene reaches Happy and Ecstatic. */
const LUXURIES = ['WINE', 'COTTON', 'DYES', 'INCENSE', 'SILVER', 'JADE'] as const;
function giveLuxuries(state: GameState, city: City, n: number): void {
  const ring = neighbors(state.map, state.map.tiles[city.centerIndex]);
  for (let i = 0; i < n; i++) {
    const t = ring[i];
    setTileOwner(t, 0, city.id);
    t.resource = LUXURIES[i];
    t.improvement = RESOURCES[LUXURIES[i]].improvement;
    t.feature = null;
  }
}

describe('the Scottish Enlightenment', () => {
  it('pays a happy city 5% more Science and Production, an ecstatic one 10%', () => {
    expect(HAPPY_YIELD_ROWS.length).toBe(4);
    const yieldsOf = (row: number, luxuries: number) => {
      const state = sceneAs(row);
      const city = settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
      city.population = 6; // needs 2 Amenities, so the luxuries decide the tier
      city.buildings.push('LIBRARY'); // something to scale
      giveLuxuries(state, city, luxuries);
      const st = computeCityStats(state, city);
      return { tier: st.amenities.tier.name, science: st.total.science, production: st.total.production, gold: st.total.gold };
    };
    const contentScot = yieldsOf(seatRow('SCOTLAND'), 1);
    const contentPlain = yieldsOf(seatRow('AMERICA'), 1);
    expect(contentScot.tier).toBe('Content');
    expect(contentScot.science).toBeCloseTo(contentPlain.science, 9);
    const happyScot = yieldsOf(seatRow('SCOTLAND'), 2);
    const happyPlain = yieldsOf(seatRow('AMERICA'), 2);
    expect(happyScot.tier).toBe('Happy');
    expect(happyScot.science).toBeCloseTo(happyPlain.science * 1.05, 9);
    expect(happyScot.production).toBeCloseTo(happyPlain.production * 1.05, 9);
    expect(happyScot.gold).toBeCloseTo(happyPlain.gold, 9); // an unnamed yield is untouched
    const ecstaticScot = yieldsOf(seatRow('SCOTLAND'), 4);
    const ecstaticPlain = yieldsOf(seatRow('AMERICA'), 4);
    expect(ecstaticScot.tier).toBe('Ecstatic');
    expect(ecstaticScot.science).toBeCloseTo(ecstaticPlain.science * 1.1, 9);
  });

  it('pays a happy city a Great Scientist point per Campus, doubled while ecstatic', () => {
    expect(HAPPY_GPP_ROWS.length).toBe(4);
    const pointsOf = (row: number, luxuries: number): number => {
      const state = sceneAs(row);
      const city = settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
      city.population = 6; // needs 2 Amenities, so the luxuries decide the tier
      const t = neighbors(state.map, state.map.tiles[city.centerIndex])[5];
      setTileOwner(t, 0, city.id);
      t.district = 'CAMPUS';
      t.districtComplete = true;
      city.districts.push({ type: 'CAMPUS', tileIndex: t.index } as City['districts'][number]);
      giveLuxuries(state, city, luxuries);
      return greatPersonPointsPerTurn(state, 0).SCIENTIST;
    };
    const plain = pointsOf(seatRow('AMERICA'), 2);
    expect(pointsOf(seatRow('SCOTLAND'), 2)).toBe(plain + 1);
    expect(pointsOf(seatRow('SCOTLAND'), 4)).toBe(plain + 2);
    expect(pointsOf(seatRow('SCOTLAND'), 1)).toBe(plain); // Content pays nothing
  });
});

describe('the roster government slots', () => {
  it("Plato's Republic adds a Wildcard, the Holy Roman Emperor a Military", () => {
    expect(POLICY_SLOT_ROWS.length).toBe(2);
    const greece = sceneAs(seatRow('GREECE'));
    expect(wonderExtraSlots(greece, 0).wildcard).toBe(1);
    expect(wonderExtraSlots(greece, 0).military).toBe(0);
    expect(wonderExtraSlots(greece, 1).wildcard).toBe(0);
    const germany = sceneAs(leaderRow('BARBAROSSA'));
    expect(wonderExtraSlots(germany, 0).military).toBe(1);
    expect(wonderExtraSlots(germany, 0).wildcard).toBe(0);
  });
});

describe('what a kill pays', () => {
  it("Gorgo takes Culture and Tamar Faith, half the defeated unit's strength", () => {
    expect(POST_COMBAT_YIELD_ROWS.length).toBe(2);
    const half = Math.floor((UNITS.WARRIOR.combat ?? 0) / 2);
    expect(half).toBeGreaterThan(0);
    const gorgo = sceneAs(leaderRow('GORGO'));
    const before = gorgo.seats[0].research.civicProgress;
    unitKillEvent(gorgo, 0, { type: 'WARRIOR' }, { type: 'WARRIOR', seat: 1 });
    expect(gorgo.seats[0].research.civicProgress).toBe(before + half);
    // a BARBARIAN victim pays too
    unitKillEvent(gorgo, 0, { type: 'WARRIOR' }, { type: 'WARRIOR', seat: 200 });
    expect(gorgo.seats[0].research.civicProgress).toBe(before + 2 * half);
    const tamar = sceneAs(leaderRow('TAMAR'));
    const faith = tamar.seats[0].faith;
    unitKillEvent(tamar, 0, { type: 'WARRIOR' }, { type: 'WARRIOR', seat: 1 });
    expect(tamar.seats[0].faith).toBe(faith + half);
    // a plain seat banks nothing, and a CIVILIAN victim has no strength
    const plain = sceneAs(seatRow('AMERICA'));
    const p0 = plain.seats[0].research.civicProgress;
    unitKillEvent(plain, 0, { type: 'WARRIOR' }, { type: 'WARRIOR', seat: 1 });
    expect(plain.seats[0].research.civicProgress).toBe(p0);
    const g2 = gorgo.seats[0].research.civicProgress;
    unitKillEvent(gorgo, 0, { type: 'WARRIOR' }, { type: 'SETTLER', seat: 1 });
    expect(gorgo.seats[0].research.civicProgress).toBe(g2);
  });

  it('Thermopylae: +1 Combat Strength for every Military policy slotted', () => {
    const state = sceneAs(leaderRow('GORGO'));
    state.unitsMode = true;
    settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
    const u = spawnUnit(state, 'WARRIOR', tileAtCoords(state.map, 5, 5).index, 0)!;
    expect(getModifiers(state, 0).militaryPolicies).toBe(0);
    expect(rosterCS(state, u, 1, 100, false)).toBe(0);
    // the first government's military slots fill as the cards unlock
    grantCivics(state, 'CODE_OF_LAWS', 'FOREIGN_TRADE', 'CRAFTSMANSHIP', 'MILITARY_TRADITION');
    const n = getModifiers(state, 0).militaryPolicies;
    expect(n).toBeGreaterThan(0);
    expect(rosterCS(state, u, 1, 100, false)).toBe(n);
    const plain = sceneAs(seatRow('AMERICA'));
    plain.unitsMode = true;
    grantCivics(plain, 'CODE_OF_LAWS', 'FOREIGN_TRADE', 'CRAFTSMANSHIP', 'MILITARY_TRADITION');
    const w = spawnUnit(plain, 'WARRIOR', tileAtCoords(plain.map, 5, 5).index, 0)!;
    expect(rosterCS(plain, w, 1, 100, false)).toBe(0);
  });
});

describe('the Amazon', () => {
  it("gives Brazil's four districts a rainforest adjacency, and nobody else's", () => {
    expect(DISTRICT_ADJ_ROWS.filter((r) => r.source === 'RAINFOREST').length).toBe(4);
    const adjOf = (row: number, type: 'CAMPUS' | 'HARBOR'): number => {
      const state = sceneAs(row);
      const city = settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
      const site = neighbors(state.map, state.map.tiles[city.centerIndex])[0];
      setTileOwner(site, 0, city.id);
      let n = 0;
      for (const t of neighbors(state.map, site) as Tile[]) {
        if (t.index === city.centerIndex) continue;
        t.feature = 'RAINFOREST';
        n += 1;
      }
      const add = getModifiers(state, 0).districtAdjacencyAdd[type] ?? [];
      return districtAdjacency(state.map, site, type, add) - districtAdjacency(state.map, site, type, []);
    };
    expect(adjOf(seatRow('BRAZIL'), 'CAMPUS')).toBeGreaterThan(0);
    expect(adjOf(seatRow('AMERICA'), 'CAMPUS')).toBe(0);
    expect(adjOf(seatRow('BRAZIL'), 'HARBOR')).toBe(0); // an unnamed district
  });
});

File: tests/gpu/seat_rows_test.py
"""THE SEAT'S ROSTER ROWS — the GPU half.

    npm run seed && npm run export        # (once) writes seeder/worlds/
    python tests/gpu/seat_rows_test.py

The TS twin is tests/cpu/seats/seat-rows.test.ts.

CIV6 (the install's TraitModifiers): the Scottish Enlightenment's happiness
percentages and Great Person points, the government slot of Plato's
Republic and the Holy Roman Emperor, the Culture and Faith a kill pays
Gorgo and Tamar, Thermopylae's per-policy strength, and the Amazon's
rainforest adjacency.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
sys.path.insert(0, str(Path(__file__).resolve().parent))
from core import BatchSim, load_rules, load_fixture, fixture_paths
from warmup import settle_all

B0 = 0
RULES = json.loads((Path(__file__).resolve().parent.parent.parent
                    / "seeder" / "worlds" / "rules.json").read_text())
UNITS = [u["id"] for u in RULES["units"]]
BARB_SEAT = 200


def play(sim, row: int, name):
    if name is None:
        sim.row_civ[0, row] = -1
        sim.row_leader[0, row] = -1
    else:
        ci = sim._civ_ids.index(name)
        sim.row_civ[0, row] = ci
        sim.row_leader[0, row] = sim._pair_civ.index(ci)
    sim._eff_version += 1
    sim._gen_ver += 1
    sim._bldg_version += 1


def lead(sim, row: int, civ: str, leader: str) -> None:
    """Seat a civilization by one NAMED leader — a civilization with two."""
    play(sim, row, civ)
    sim.row_leader[0, row] = sim._leader_idx(leader)
    sim._eff_version += 1


def fresh(rules, path) -> BatchSim:
    sim = BatchSim([load_fixture(path)], rules, device="cpu", dtype=torch.float64)
    for r, name in enumerate(("ROME", "EGYPT", "NORWAY")):
        play(sim, r, name)
    return settle_all(sim)


def T(*xs) -> torch.Tensor:
    return torch.tensor(list(xs), dtype=torch.long)


# ---------------------------------------------------------------------------


def test_wire(rules, path) -> None:
    sim = fresh(rules, path)
    assert len(sim._happy_yield_rows) == 4 and len(sim._happy_gpp_rows) == 4
    assert len(sim._policy_slot_rows) == 2 and len(sim._post_combat_yield_rows) == 2
    assert sum(1 for r in sim._district_adj_rows if r[4] == 2) == 4, "the Amazon's four"
    assert sum(1 for r in sim._combat_cs_rows if r[5] == 1) == 1, "Thermopylae's magnitude"
    print("  1 wire OK — 4 + 4 + 2 + 2 rows, 4 feature adjacencies, 1 per-policy row")


def test_policy_slots(rules, path) -> None:
    sim = fresh(rules, path)
    play(sim, 0, "ROME")
    base = sim._wonder_extra_slots(0)[B0].tolist()
    play(sim, 0, "GREECE")
    got = sim._wonder_extra_slots(0)[B0].tolist()
    assert got[3] == base[3] + 1 and got[0] == base[0], "Plato's Republic — a WILDCARD"
    lead(sim, 0, "GERMANY", "BARBAROSSA")
    got = sim._wonder_extra_slots(0)[B0].tolist()
    assert got[0] == base[0] + 1 and got[3] == base[3], "the Holy Roman Emperor — a MILITARY"
    play(sim, 0, "ROME")
    assert sim._wonder_extra_slots(0)[B0].tolist() == base, "the slot outlived the roster row"
    print("  2 policy slots OK — a Wildcard for Greece, a Military for Barbarossa")


def test_kill_yields(rules, path) -> None:
    warrior = UNITS.index("WARRIOR")
    settler = UNITS.index("SETTLER")
    half = int(RULES["units"][warrior]["combat"]) // 2
    assert half > 0

    def killed_as(civ: str, leader: str, vict: int, barb: bool) -> tuple[int, int]:
        sim = fresh(rules, path)
        lead(sim, 0, civ, leader)
        c0 = float(sim.civ_civic_prog[B0, 0])
        f0 = float(sim.civ_faith[B0, 0])
        sim._unit_kill_event(T(0), T(vict), torch.tensor([barb]), torch.tensor([True]))
        return int(float(sim.civ_civic_prog[B0, 0]) - c0), int(float(sim.civ_faith[B0, 0]) - f0)

    assert killed_as("GREECE", "GORGO", warrior, False) == (half, 0), "Thermopylae's Culture"
    assert killed_as("GREECE", "GORGO", warrior, True) == (half, 0), "a barbarian victim pays too"
    assert killed_as("GREECE", "GORGO", settler, False) == (0, 0), "a civilian has no strength"
    assert killed_as("GEORGIA", "TAMAR", warrior, False) == (0, half), "Tamar's Faith"
    assert killed_as("ROME", "TRAJAN", warrior, False) == (0, 0), "a plain seat banked a kill"
    print(f"  3 kill yields OK — {half} Culture for Gorgo, {half} Faith for Tamar")


def test_thermopylae(rules, path) -> None:
    sim = fresh(rules, path)
    lead(sim, 0, "GREECE", "GORGO")
    warrior = UNITS.index("WARRIOR")
    land = next(t for t in range(sim.T) if not bool(sim.water[B0, t]) and bool(sim.passable[B0, t]))
    hp = torch.tensor([100.0], dtype=sim.unit_hp.dtype)

    def cs() -> int:
        return int(sim._roster_cs(T(0), T(warrior), T(land), T(1), hp, False)[B0])

    assert int(sim._military_policies(T(0))[B0]) == 0 and cs() == 0, "a card-less seat"
    civics = [c["id"] for c in RULES["civics"]]
    for name in ("CODE_OF_LAWS", "FOREIGN_TRADE", "CRAFTSMANSHIP", "MILITARY_TRADITION"):
        sim.civ_civics[B0, 0, civics.index(name)] = True
    sim._eff_version += 1
    sim._gov_pol_cache.clear()
    n = int(sim._military_policies(T(0))[B0])
    assert n > 0, "no Military policy slotted — the scene measures nothing"
    assert cs() == n, f"the per-policy magnitude read {cs()} against {n} slotted"
    play(sim, 0, "ROME")
    sim._gov_pol_cache.clear()
    assert cs() == 0, "the magnitude outlived Gorgo"
    print(f"  4 Thermopylae OK — +1 per slotted Military policy ({n} slotted)")


def test_amazon(rules, path) -> None:
    sim = fresh(rules, path)
    campus = next(i for i, d in enumerate(sim.districts_cat) if d["id"] == "CAMPUS")
    harbor = next(i for i, d in enumerate(sim.districts_cat) if d["id"] == "HARBOR")
    # a tile of this row's own ground with rainforest around it
    t = next(t for t in range(sim.T) if int(sim.tile_seat[B0, t]) == 0 and not bool(sim.water[B0, t]))
    ring = [int(x) for x in sim.neigh[t].tolist() if x >= 0]
    for x in ring:
        sim.feat_id[B0, x] = sim._rainforest_fid
        sim.feat_stripped[B0, x] = False
    sim._eff_version += 1
    play(sim, 0, "ROME")
    base_c = float(sim._district_adj_floor(campus)[B0, t])
    base_h = float(sim._district_adj_floor(harbor)[B0, t])
    play(sim, 0, "BRAZIL")
    got_c = float(sim._district_adj_floor(campus)[B0, t])
    got_h = float(sim._district_adj_floor(harbor)[B0, t])
    assert got_c == base_c + len(ring), f"the Amazon paid {got_c - base_c} for {len(ring)} rainforest"
    assert got_h == base_h, "the Amazon paid a Harbor"
    print(f"  5 the Amazon OK — +1 per adjacent Rainforest on four districts ({len(ring)} around)")


def test_happy_rows(rules, path) -> None:
    """The happiness rows read the tier `_seat_amenity` decides."""
    sim = fresh(rules, path)
    tiers = sim.rules.amenity_tiers
    ecstatic, happy = 0, 1
    assert tiers[happy][0] == 1 and tiers[ecstatic][0] == 3, "the tier order moved"

    def give_luxuries(sim, n: int) -> None:
        """Improved luxuries on this row's own ground — the amenity source
        `luxuryAmenities` ranks, so the tier follows the count."""
        seen: set[int] = set()
        for t in range(sim.T):
            if len(seen) >= n:
                break
            k = int(sim.lux_id[B0, t])
            if k < 0 or k in seen or int(sim.tile_seat[B0, t]) != 0:
                continue
            sim.improvement[B0, t] = int(sim.lux_req[B0, t])
            sim.pillaged[B0, t] = False
            seen.add(k)
        sim._eff_version += 1

    def sci_and_tier(name, pop: int, lux_add: int) -> tuple[float, int]:
        s2 = fresh(rules, path)
        play(s2, 0, name)
        s2.city_pop[B0, 0, 0] = pop
        give_luxuries(s2, lux_add)
        tier = int(s2._seat_amenity(0)[0][B0, 0])
        sci = float(s2._seat_city_walk(0, amen_yf=s2._seat_amenity(0)[2])[B0, 0, 3])
        return sci, tier

    seen_tier = False
    for pop, lux in ((6, 0), (6, 2), (6, 4), (1, 0), (1, 4)):
        base, t_base = sci_and_tier("ROME", pop, lux)
        scot, t_scot = sci_and_tier("SCOTLAND", pop, lux)
        assert t_base == t_scot, "the roster row moved the TIER itself"
        want = 1.05 if t_base == happy else 1.1 if t_base == ecstatic else 1.0
        assert abs(scot - base * want) < 1e-9, f"tier {t_base} paid {scot / base if base else 0}"
        if t_base in (happy, ecstatic):
            seen_tier = True
    assert seen_tier, "no scene reached Happy or Ecstatic — the rows went unmeasured"
    print("  6 happiness rows OK — 5% Happy and 10% Ecstatic on Science, Content untouched")


def main() -> int:
    rules = load_rules()
    path = fixture_paths()[0]
    test_wire(rules, path)
    test_policy_slots(rules, path)
    test_kill_yields(rules, path)
    test_thermopylae(rules, path)
    test_amazon(rules, path)
    test_happy_rows(rules, path)
    print("BATTERY OK seat_rows")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
